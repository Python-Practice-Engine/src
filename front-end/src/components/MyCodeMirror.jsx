import React from 'react';

// Auxiliary helper
const mirror = require('codemirror');

// Component defining the CodeMirror instance to be used by the IDE
class MyCodeMirror extends React.Component {
  // Load in the relevant third-party plugin
  componentDidMount() {
    // Initialize CodeMirror
    const options = {
      mode: 'python',
      theme: 'mdn-like',
      lineNumbers: true,
      lineWrapping: true,
    };
    if (document.getElementById('code-input') !== null) {
      this.codeMirror = mirror.fromTextArea(
        document.getElementById('code-input'),
        options,
      );
      this.codeMirror.setSize('100%', '100%');
      this.codeMirror.on('change', this.onEdit);
    }
  }

  // Unload third-party plugin after finished
  componentWillUnmount() {
    // initialize CodeMirror once done
    if (this.codeMirror) {
      this.codeMirror.toTextArea();
      this.codeMirror = null;
    }
  }

  // Update the Ui when user changes something in the code editor
  onEdit = (editor) => {
    // Modifies editor
    const change = this.props;
    document.getElementById('code-input').value = editor.getValue();
    if (change.onChange) {
      change.onChange(editor.getValue());
    }
  };

  // Display the UI overlay to the IDE
  render() {
    return (
      <textarea
        id="code-input"
        name="code-input"
        margin="0"
        onChange={this.onEdit}
      />
    );
  }
}

export default MyCodeMirror;
