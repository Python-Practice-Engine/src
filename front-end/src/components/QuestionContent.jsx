import React from 'react';

// React component library imports
import {
  Typography,
  Tag,
  Space,
} from 'antd';

const { Title } = Typography;

/*
  This component renders the content of each question. The content includes:
  the question title, the questions itself, the question's category.
*/
class QuestionContent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div>
        <Title
          level={3}
        >
          {this.props.question.title}
        </Title>
        <Space>
          <Tag color="warning">{this.props.concept.name}</Tag>
        </Space>
        <p className="questionDescription">
          {this.props.question.description}
        </p>
      </div>

    );
  }
}

export default QuestionContent;
