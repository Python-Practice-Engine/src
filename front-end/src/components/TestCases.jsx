import React from 'react';

// React component library imports
import {
  List,
  Typography,
  Tag,
  Card,
} from 'antd';

// Variable declarations
const { Title } = Typography;
const testCaseTitle = 'Test Case #';

// Used to render the test-cases table in the IDE page

function TestCase(props) {
  // eslint-disable-next-line prefer-destructuring
  const number = props.item.number;
  const passed = props.item.passed === undefined ? null : props.item.passed;
  if (passed === null) {
    return (
      <div>
        <Tag
          style={{
            display: 'inline-flex',
            alignItems: 'center',
          }}
          id={number}
          color="#1890ff"
        >
          Click submit to run tests
        </Tag>
      </div>
    );
  }
  if (passed === 0) {
    return (
      <div>
        <Tag
          style={{
            display: 'inline-flex',
            alignItems: 'center',
          }}
          id={number}
          color="red"
        >
          Failed
        </Tag>
      </div>
    );
  }
  if (passed === 1) {
    return (
      <div>
        <Tag
          style={{
            display: 'inline-flex',
            alignItems: 'center',
          }}
          id={number}
          color="green"
        >
          Passed
        </Tag>
      </div>
    );
  }
}
class TestCases extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  // Display the test-cases table on IDE page
  render() {
    return (
      <div className="testCases">
        <Card>
          <Title
            level={3}
          >
            Test Cases
          </Title>
          <List
            style={{
              width: '105%',
              marginLeft: '-12px',
            }}
            size="large"
            bordered
            dataSource={this.props.testCases}
            renderItem={
              (item) => (
                <Card
                  title={testCaseTitle + item.number}
                  className="test-title"
                  bordered={false}
                >
                  <TestCase item={item} />
                  {item.test}
                </Card>
              )
            }
          />
        </Card>
      </div>
    );
  }
}

export default TestCases;
