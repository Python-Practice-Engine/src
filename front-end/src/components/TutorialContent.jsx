import React from 'react';

// React component library imports
import {
  Typography,
  Space,
} from 'antd';

const { Title } = Typography;

// populating tutorial tab for main page
function TutorialContent(props) {
  const content = props.contents;
  return (
    <div>
      <Title
        level={3}
      >
        {`${content.name} ${content.level}`}
      </Title>
      <Space />
      <p className="questionDescription">
        {content.tutorial}
      </p>
    </div>
  );
}

export default TutorialContent;
