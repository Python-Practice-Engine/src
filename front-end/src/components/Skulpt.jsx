import React from 'react';
import Axios from 'axios';

// React component library imports
import {
  Input,
  Button,
  Row,
  Typography,
  Card,
} from 'antd';
import {
  LeftOutlined,
  RightOutlined,
} from '@ant-design/icons';

// Imports for React routing
import {
  Link,
  HashRouter,
  NavLink,
} from 'react-router-dom';

// Code mirror imports
import 'codemirror/mode/python/python';

// Personal component imports
import MyCodeMirror from './MyCodeMirror';

import AccountContext from './Account';

const { Title } = Typography;
/* File consists of auxiliary functions for IDE rendering */
// Used by third-party plugin to compute sent code
function builtinRead(x) {
  if (
    window.Sk.builtinFiles === undefined
    || window.Sk.builtinFiles.files[x] === undefined
  ) {
    throw new Error(`File not found: '${x}'`);
  }
  return window.Sk.builtinFiles.files[x];
}

// Component to be used by the IDE to execute any user created code
class Skulpt extends React.Component {
  // eslint-disable-next-line react/static-property-placement
  static contextType = AccountContext;

  constructor(props) {
    super(props);
    this.execute = this.execute.bind(this);
    this.submit = this.submit.bind(this);
    this.outf = this.outf.bind(this);
    this.state = {
      output: '',
    };
  }

  // loading relvant third-party plugin
  componentDidMount() {
    const script = document.createElement('script');
    script.src = '../skulpt.js';
    script.async = true;
  }

  // Used to output computed code
  outf = (text) => {
    const mypre = document.getElementById('code-output');
    const output = document.getElementById('output-area');
    output.style.display = 'block';
    this.setState((state) => ({
      output: state.output + text,
    }));
    mypre.innerHTML += text;
  }

  mark = async (i, tests, mark) => {
    const marked = await Axios.post(`http://localhost:3001/mark_test_case/${this.context.user}/${tests[i].test_case_id}/${mark}/${this.props.questionId}/${this.props.conceptId}`);
    // const nextQuestionData = await marked.data === true ? Axios.get(`http://localhost:3001/next_question/${this.context.user}`) : false;
    if (await marked.data === true) {
      const nextQuestionData = await Axios.get(`http://localhost:3001/next_question/${this.context.user}`);
      const id = await nextQuestionData.data[0].id;
      this.props.nextQuestion(await id);
    }
  }

  // Auxiliary function that allows the user to run
  // the given questions test-cases
  submit() {
    // Variables declarations
    const tests = this.props.testCases;
    const codeOutput = document.getElementById('code-input');

    if (codeOutput) {
      // Loop through test-cases for given question
      // eslint-disable-next-line spaced-comment
      for (let i = 0; i < tests.length; i += 1) {
        // Append tests individually to code then execute
        const codeTR = tests[i].code;
        let expect = tests[i].expected;
        const prog = codeOutput.value + codeTR.replaceAll('\\', '');

        // Auxiliary preperations for Skulpt
        this.setState({ output: '' });
        window.Sk.python3 = true;
        window.Sk.pre = 'output';
        window.Sk.configure({
          output: this.outf,
          read: builtinRead,
          killableWhile: true,
          yieldLimit: 100,
          execLimit: 5000,
          timeoutMsg: 'your program timed out',
          python3: true,
        });

        // Feed code into Skulpt to execute
        const myPromise = window.Sk.misceval.asyncToPromise(
          () => window.Sk.importMainWithBody(
            '<stdin>',
            true,
            prog,
            true,
          ),
        );
        // Code executes
        myPromise.then(() => {
          // Display 'Passed' tag if test passes
          const scrollingElement = (
            document.scrollingElement || document.body
          );
          scrollingElement.scrollTop = scrollingElement.scrollHeight;

          expect += '\n';
          if (this.state.output === expect) {
            this.props.updateTestCases(i, 1);
            this.mark(i, tests, 1);
          } else {
            this.props.updateTestCases(i, 0);
            this.mark(i, tests, 0);
          }
          console.log('success');
        },
        // Error in code, output error message
        (error) => {
          const errMsg = error.toString();
          this.outf(errMsg);
        });
      }
    }
  }

  // Auxiliary function that allows the user to run their code without
  // running the questions tests
  execute() {
    // Variable declarations
    const codeOutput = document.getElementById('code-input');

    if (codeOutput) {
      // DOM elements to be used by Skulpt
      const prog = codeOutput.value;

      // Auxiliary preperations for Skulpt
      this.setState({ output: '' });
      window.Sk.python3 = true;
      window.Sk.pre = 'output';
      window.Sk.configure({
        output: this.outf,
        read: builtinRead,
        killableWhile: true,
        yieldLimit: 100,
        execLimit: 5000,
        timeoutMsg: 'your program timed out',
        python3: true,
      });

      // Feed code into Skulpt to execute
      const myPromise = window.Sk.misceval.asyncToPromise(
        () => window.Sk.importMainWithBody(
          '<stdin>',
          true,
          prog,
          true,
        ),
      );
      // Code executes properly
      myPromise.then(() => {
        const scrollingElement = (
          document.scrollingElement || document.body
        );
        scrollingElement.scrollTop = scrollingElement.scrollHeight;
        console.log('success');
        return 'success';
      },
      // Error in code, output error message
      (error) => {
        const errMsg = error.toString();
        this.outf(errMsg);
      });
    }
  }

  // Used to Display the IDE portion of the IDE page
  render() {
    return (
      <div className="ide-card">
        <div className="text-editor">
          <Input.Group compact size="large" className="mock-input">
            <MyCodeMirror />
          </Input.Group>
        </div>
        <HashRouter>
          <Row
            type="flex"
            align="middle"
            className="submit-btns"
            justify="space-between"
          >
            <NavLink
              tag={Link}
              to={
                `/user/${this.context.user}/question/${
                  this.props.easierQuestionId
                }`
              }
            >
              <Button
                type="primary"
                size="large"
                icon={<LeftOutlined />}
              />
            </NavLink>
            <NavLink
              tag={Link}
              to={
                `/user/${this.context.user}/question/${
                  this.props.nextQuestionId
                }`
              }
            >
              <Button
                type="primary"
                size="large"
                icon={<RightOutlined />}
              />
            </NavLink>
            <Button
              id="Run"
              type="primary"
              size="large"
              onClick={this.execute}
              className="run-btn"
            >
              Run
            </Button>
            <Button
              id="Submit"
              size="large"
              onClick={this.submit}
            >
              Submit
            </Button>
          </Row>
          <Row>
            <Title
              level={5}
              style={{ marginLeft: '2.25%', fontWeight: 'normal' }}
            >
              Difficulty
            </Title>
          </Row>
        </HashRouter>
        <div id="output-area">
          <Card
            type="inner"
            title="Output"
            style={{ padding: '0', margin: '0' }}
          >
            <textarea
              value={this.state.output}
              readOnly="{true}?"
              className="output"
              id="code-output"
              name="output"
            />
          </Card>
        </div>
      </div>
    );
  }
}

export default Skulpt;
