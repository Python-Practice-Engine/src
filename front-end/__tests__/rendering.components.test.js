import React from 'react';
import renderer from 'react-test-renderer';
import TutorialContent from '../src/components/TutorialContent';
import TestCases from '../src/components/TestCases';
import MyCodeMirror from '../src/components/MyCodeMirror';
import QuestionContent from '../src/components/QuestionContent';
import Skulpt from '../src/components/Skulpt';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { mount, configure } from 'enzyme';

configure({adapter: new Adapter()});

// Workaround source: https://stackoverflow.com/questions/39830580/jest-test-fails-typeerror-window-matchmedia-is-not-a-function
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Tests that the TutotiralContent component properly renders the relvant HTML
it('renders the tutorial content correctly', () => {
const tree = renderer
  .create(<TutorialContent contents={
      {
          name: "tut1",
          description: "description"
      }

  } >

  </TutorialContent>)
  .toJSON();
expect(tree).toMatchSnapshot();
});

// Tests that the TestCases component properly renders the relvant HTML
// As well as properly displaying retrieved testcase information
it('renders the test cases correctly', () => {
  const tree = renderer
    .create(<TestCases testCases={
        [
            {
                test: "test 1",
                TCid: 1,
            },
            {
              test: "test 2",
              TCid: 2,
          },
        ]
    }></TestCases>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// Tests that the MyCodeMirror component properly renders the relvant HTML
it('renders the code mirror correctly', () => {
const tree = renderer
  .create(<MyCodeMirror onChange={true}></MyCodeMirror>)
  .toJSON();
expect(tree).toMatchSnapshot();
});

// Tests that the execute() method of the
// Skulpt third-party library is properly called
it('skulpt execute() interperts user python code', async () => {
const tree = mount(<Skulpt></Skulpt>);
const instance = tree.instance();
const compute = jest.spyOn(instance, 'execute');

instance.execute();

expect(compute).toHaveBeenCalledTimes(1);
});

// Tests that the Run button is properly connected to the execute() method
it('run call execute() and computes code', async () => {
const tree = mount(<Skulpt></Skulpt>);

tree.find("Button#Run").prop('onClick');

expect(tree.find("#output-area")).toHaveLength(1);
});

// Tests that the execute() method of the Skulpt third-party library 
// runs with no errors
it('tests that skulpt execute() throws no errors', async () => {
const tree = mount(<Skulpt></Skulpt>);
const instance = tree.instance();
const compute = jest.spyOn(instance, 'execute');
instance.execute();

expect(compute).toHaveReturned();
});

// Tests that the submit() method of the Skulpt third-party 
// library is properly called
it('skulpt submit() interperts user python code', async () => {
const tree = mount(<Skulpt></Skulpt>);
const instance = tree.instance();
const compute = jest.spyOn(instance, 'submit');
instance.submit();

expect(compute).toHaveBeenCalledTimes(1);
});

// Tests that the Submit button is connected to the submit() method
it('submit calls submit() and computes code', async () => {
const tree = mount(<Skulpt></Skulpt>);

tree.find("Button#Submit").prop('onClick');

expect(tree.find("#output-area")).toHaveLength(1);
});

// Tests that the submit() method of the Skulpt third-party 
// library runs with no errors
it('tests that skulpt submit() throws no errors', async () => {
const tree = mount(<Skulpt></Skulpt>);
const instance = tree.instance();
const compute = jest.spyOn(instance, 'submit');
instance.submit();

expect(compute).toHaveReturned();
});

// Tests that the QuestionContent component properly renders the relvant HTML
// As well as properly displaying retrieved question details
it('renders the question content correctly', () => {
const tree = renderer
  .create(<QuestionContent contents={
      {
          name: "q1",
          question: "why is the sky blue?",
          tags: "easy",
      }
  } >
  </QuestionContent>)
  .toJSON();
expect(tree).toMatchSnapshot();
});
