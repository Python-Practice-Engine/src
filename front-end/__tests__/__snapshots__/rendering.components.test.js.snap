// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the code mirror correctly 1`] = `
<textarea
  id="code-input"
  margin="0"
  name="code-input"
  onChange={[Function]}
  value="#Your Code Here"
/>
`;

exports[`renders the question content correctly 1`] = `
<div>
  <h3
    className="ant-typography"
    style={
      Object {
        "WebkitLineClamp": null,
      }
    }
  >
    q1
  </h3>
  <div
    className="ant-space ant-space-horizontal ant-space-align-center"
    style={Object {}}
  >
    <div
      className="ant-space-item"
      style={Object {}}
    >
      <span
        className="ant-tag ant-tag-warning"
        style={
          Object {
            "backgroundColor": undefined,
          }
        }
      >
        easy
      </span>
    </div>
  </div>
  <p
    className="questionDescription"
  >
    why is the sky blue?
  </p>
</div>
`;

exports[`renders the test cases correctly 1`] = `
<div
  className="testCases"
>
  <div
    className="ant-card ant-card-bordered"
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <h3
        className="ant-typography"
        style={
          Object {
            "WebkitLineClamp": null,
          }
        }
      >
        Test Cases
      </h3>
      <div
        className="ant-list ant-list-lg ant-list-split ant-list-bordered"
        style={
          Object {
            "marginLeft": "-12px",
            "width": "105%",
          }
        }
      >
        <div
          className="ant-spin-nested-loading"
        >
          <div
            className="ant-spin-container"
          >
            <ul
              className="ant-list-items"
            >
              <div
                className="ant-card test-title"
              >
                <div
                  className="ant-card-head"
                  style={Object {}}
                >
                  <div
                    className="ant-card-head-wrapper"
                  >
                    <div
                      className="ant-card-head-title"
                    >
                      Test Case #1
                    </div>
                  </div>
                </div>
                <div
                  className="ant-card-body"
                  style={Object {}}
                >
                  <span
                    className="ant-tag ant-tag-success"
                    id={1}
                    style={
                      Object {
                        "backgroundColor": undefined,
                        "display": "none",
                      }
                    }
                  >
                    Passed
                  </span>
                  test 1
                </div>
              </div>
              <div
                className="ant-card test-title"
              >
                <div
                  className="ant-card-head"
                  style={Object {}}
                >
                  <div
                    className="ant-card-head-wrapper"
                  >
                    <div
                      className="ant-card-head-title"
                    >
                      Test Case #2
                    </div>
                  </div>
                </div>
                <div
                  className="ant-card-body"
                  style={Object {}}
                >
                  <span
                    className="ant-tag ant-tag-success"
                    id={2}
                    style={
                      Object {
                        "backgroundColor": undefined,
                        "display": "none",
                      }
                    }
                  >
                    Passed
                  </span>
                  test 2
                </div>
              </div>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the tutorial content correctly 1`] = `
<div>
  <h3
    className="ant-typography"
    style={
      Object {
        "WebkitLineClamp": null,
      }
    }
  >
    tut1
  </h3>
  <p
    className="questionDescription"
  >
    description
  </p>
</div>
`;
